src/main/java/org/opentripplanner/updater/package.md

```diff
@@ -0,0 +1,104 @@
+
+This is essentially a multi-version snapshot concurrency control system, inspired by widely used database engines (and in fact informed by books on transactional database design). The end result is a system where 1) writing operations are simple to reason about and cannot conflict because only one write happens at a time; 2) multiple read operations (including routing requests) can occur concurrently; 3) read operations do not need to pause while writes are happening; 4) read operations see only fully completed write operations, never partial writes; and 5) each read operation sees a consistent, unchanging view of the transit data.
+
+An important characteristic of this approach is that _no locking is necessary_. However, some form of synchronization is used during the buffer swap operation to impose a consistent view of the whole data structure via a happens-before relationship as defined by the Java memory model. While pointers to objects can be handed between threads with no read-tearing of the pointer itself, there is no guarantee that the web of objects pointed to will be consistent without some explicit synchronization at the hand-off.
+
+Arguably the process of creating an immutable live snapshot (and a corresponding new writable buffer) should be handled by a GraphWriterRunnable on the single graph updater thread. This would serve to defer any queued modifications until the new buffer is in place, without introducing any further locking mechanisms.
+
+## Design Considerations
+
+This section summarizes the rationale behind some of the design decisions.
+
+### Realtime is Highly Concurrent
+
+An OTP instance can have multiple sources of realtime data at once. In some cases the transit data includes several feeds of scheduled data from different providers, with one or more types of realtime updates for those different feeds.
+
+In a large production system, ideally all the scheduled data would be integrated into a single data source with a unified ID namespace, and all the realtime data would also be integrated into a single data source with an exactly matching namespace. This would be the responsibility of a separate piece of software outside (upstream of) OTP. In practice, such an upstream data integration system does not exist in many deployments and OTP must manage several static and realtime data sources at once. Even when data feeds are well-integrated, the different kinds of realtime (arrival time updates, vehicle positions, or text alerts) may be split across multiple feeds as described in the GTFS-RT spec, which implies polling three different files.
```

Comment:
I wasn't considering a whole-world case, where many other things about OTP also wouldn't scale. I was considering current typical OTP instances, where the largest ones are usually for single countries like Norway, Germany, or France and many deployments are much smaller.

Here I was trying to express something that I heard in past discussion: Certain responsibilities are outside the scope of OTP, and I was under the impression most of us considered a data integration pipeline to be an external service upstream of OTP. Similar to how realtime prediction is also seen as ideally outside and upstream of OTP.

If this is not the case, and integration of multiple data sources within OTP itself is considered advantageous (rather than just a convenience feature) I would like to document why and how. Is the key fact that some data sources update more frequently than others, and they update at different times? So rather than producing and ingesting the whole integrated data feed each time one of the constituent feeds changed, are you thinking of live OTP instances live-patching large chunks of their internal data? Basically, entire feeds would be replaced in the same way as realtime messages are currently applied? If this is the case we should begin documenting these ideas in the realtime working group doc.

On ID namespaces: If all the IDs across many feeds are guaranteed to be unique, in some sense they are already using a unified namespace. The "upstream integration system" I mentioned doesn't have to be merging the feeds - it could also just be data consistency and referential integrity checks for example.

Reply:
I do not think we disagree here, I am not talking about "live-patching large chunks of their internal data". If you do a single netex data source or multiple files it both comes with a set of challenges you need to deal with - one way is not better than the other - in my eyes. I also think that adding a feedScope to the ID should be an opt in feature - the id is changed. It divide the ecosystem in "downstream" and "upstream" of OTP.